Feature/issue 205 local streaming processing

chenziliang opened the pull request:
PR checklist:
- Did you run ClangFormat ?
- Did you separate headers to a different section in existing community code base ?
- Did you surround `proton: starts/ends` for new code in existing community code base ?

Please write user-readable short description of the changes:

Tumble 

```
root@t+ programs % top -cb -d 1 |
  awk '{printf("%s,%s,%s,\"%s\"\n", $1, $9, $10, $12)}' |
  ./proton local --input-format LineAsString -S "proc string" -q "select * from stream where match(proc, '^\s*\d+') settings max_block_size=1" |
  ./proton local --file - --input-format CSV -S "pid uint, cpu_util float, mem_util float, command string" \
    -q "select window_start, max_k(cpu_util, 3, command), max_k(mem_util, 3, command) from tumble(stream, now(), 1s) group by window_start settings max_block_size=1, input_format_parallel_parsing=false"
```

Global

```
top -cb -d 1 |
 awk '{printf("%s,%s,%s,%s\n", $1, $9, $10, $12)}' |
 ./proton local --input-format LineAsString -S "proc string" -q "select * from stream where match(proc, '^\s*\d+') settings max_block_size=1" | 
 ./proton local --file - --input-format CSV  -S "pid uint, cpu_util float, mem_util float, command string" \
        -q "select max_k(cpu_util, 3, command), max_k(mem_util, 3, command) from stream settings max_block_size=1, input_format_parallel_parsing=false"
```

chenziliang committed:
local streaming processing

chenziliang committed:
local streaming processing, fixes

chenziliang committed:
local streaming processing, fixes

yl-lisen commented on src/Storages/StorageFile.cpp:
```diff
@@ -481,19 +483,31 @@ class StorageFileSource : public ISource
                  std::make_shared<DataTypeLowCardinality>(std::make_shared<DataTypeString>()),
                  "_file"});
 
+        /// proton : starts
+        if (is_streaming_)
+        {
+            for (const auto * reserved_col :
+                 {&ProtonConsts::STREAMING_WINDOW_START,
+                  &ProtonConsts::STREAMING_WINDOW_END,
+                  &ProtonConsts::STREAMING_TIMESTAMP_ALIAS})
+                if (header.has(*reserved_col))
+                    header.erase(*reserved_col);
+        }
+        /// proton : ends
```
These columns have been removed in `ProxyStream`. I think we don't need this code.

chenziliang replied:
Yes they are but they are in StorageSnapshot and re-appear in StorageFile since its source column handling logic is different than other regular storage engine

yl-lisen commented on src/Storages/StorageFile.cpp:
```diff
@@ -513,6 +528,22 @@ class StorageFileSource : public ISource
+        else
+            block_for_format = storage_snapshot->metadata->getSampleBlock();
+
+        if (is_streaming)
+        {
+            for (const auto * reserved_col :
+                 {&ProtonConsts::STREAMING_WINDOW_START, &ProtonConsts::STREAMING_WINDOW_END, &ProtonConsts::STREAMING_TIMESTAMP_ALIAS})
+                if (block_for_format.has(*reserved_col))
+                    block_for_format.erase(*reserved_col);
```
These columns have been removed in `ProxyStream`. I think we don't need this code.

chenziliang committed:
More func

chenziliang committed:
format

chenziliang committed:
fix tests

yokofly committed:
Revert "fix tests"

This reverts commit 8a9c46aa52701a629d10ee07966a4fca5db0c9d2.
Reason:
Stateless tests will
iterate the `tests/queries_ported/0_stateless/01514_parallel_formatting.sql`
instead the `tests/queries/0_stateless/01514_parallel_formatting.sql`

yokofly committed:
fix test
1. 02270_stdin_with_query_or_infile_data.sh:
```cpp
// ClientBase.cpp:1102
        SelectQueryInfo query_info;
        try
        {
            auto metadata = storage->getInMemoryMetadataPtr();
            QueryPlan plan;
            storage->read(
                    plan,
                    sample.getNames(),
                    storage->getStorageSnapshot(metadata, global_context),
                    query_info,   -->>>>>>>>>>>>>>>>>>> empty info here
                    global_context,
                    {},
                    global_context->getSettingsRef().max_block_size,
                    getNumberOfPhysicalCPUCores());
```
2. 01514_parallel_formatting.sql:
ummm I add the correct mode, maybe helpful? local testing works
3. 02125_lz4_compression_bug.sh:
ummm. same above. I set the query mode to table.

yokofly committed:
Merge remote-tracking branch 'origin/develop' into feature/issue-205-local-streaming-processing

yokofly commented on tests/queries_ported/0_stateless/02125_lz4_compression_bug.sh:
```diff
@@ -9,12 +9,12 @@ for format in Native Values JSONCompactEachRow TSKV TSV CSV JSONEachRow JSONComp
     echo $format
     ${CLICKHOUSE_CLIENT} --query "DROP STREAM IF EXISTS file"
-    ${CLICKHOUSE_CLIENT} --query "CREATE STREAM file (x uint64) ENGINE = File($format, '${CLICKHOUSE_DATABASE}/data.$format.lz4')"
+    ${CLICKHOUSE_CLIENT} --query "CREATE STREAM file (x uint64) ENGINE = File($format, '${CLICKHOUSE_DATABASE}/data.$format.lz4') SETTINGS query_mode='table'"
```
incorrect modification, I will debug this tomorrow.
this script the whole result has a major difference from the expected one.
for local testing can use 
```bash
#!/usr/bin/env bash
# Tags: no-parallel, no-fasttest

for format in Native Values JSONCompactEachRow TSKV TSV CSV JSONEachRow JSONCompactEachRow JSONStringsEachRow
do
    echo $format
    ./proton-client --query "DROP STREAM IF EXISTS file"
    ./proton-client --query "CREATE STREAM file (x uint64) ENGINE = File($format, 'default/data.$format.lz4')"
    for size in 10000 100000 1000000 2500000
    do
        ./proton-client --query "TRUNCATE STREAM file"
        ./proton-client --query "INSERT INTO file SELECT * FROM numbers($size)"
        ./proton-client --query "SELECT max(x) FROM file"
    done
done

./proton-client --query "DROP STREAM file"

```

yokofly committed:
fix incorrect usage `settings query_mode='table'`

chenziliang commented on src/Storages/StorageFile.cpp:
```diff
@@ -725,7 +725,13 @@ Pipe StorageFile::read(
-    if (query_info.streaming_window_params && query_info.syntax_analyzer_result->streaming)
+    bool is_streaming = false;
+
+    if (query_info.syntax_analyzer_result) {
```
remove the `{}` and it is not conforming to our code style.

chenziliang commented on programs/local/LocalServer.cpp:
```diff
@@ -448,6 +448,23 @@ void LocalServer::connect()
 int LocalServer::main(const std::vector<std::string> & /*args*/)
 try
 {
+    const char * holloween = R"(.___,_______,_____Happy_Halloween____.
```
Let's get rid of this since Holloween already passed :)

yokofly committed:
resolve comment:
1. remove holloween specific
2. remove extra `{}`
